components/TextFields/src/ContainedInputView/MDCOutlinedTextField.m

```diff
@@ -0,0 +1,83 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCOutlinedTextField.h"
+
+#import <Foundation/Foundation.h>
+
+#import "private/MDCBaseTextField+MDCTextControl.h"
+#import "private/MDCTextControl.h"
+#import "private/MDCTextControlStyleOutlined.h"
+
+@interface MDCOutlinedTextField ()
+@end
+
+@implementation MDCOutlinedTextField
+
+#pragma mark Object Lifecycle
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonMDCOutlinedTextFieldInit];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  self = [super initWithCoder:aDecoder];
+  if (self) {
+    [self commonMDCOutlinedTextFieldInit];
+  }
+  return self;
+}
+
+- (void)commonMDCOutlinedTextFieldInit {
+  self.containerStyle = [[MDCTextControlStyleOutlined alloc] init];
+  self.borderStyle = UITextBorderStyleNone;
+}
+
+#pragma mark UITextField Overrides
+
+- (void)setBorderStyle:(UITextBorderStyle)borderStyle {
+  if (borderStyle == UITextBorderStyleNone) {
+    [super setBorderStyle:borderStyle];
+  } else {
+    NSLog(@"Setting borderStyle on MDCOutlinedTextField is not allowed.");
+    [super setBorderStyle:UITextBorderStyleNone];
+  }
+}
+
+#pragma mark Stateful Color APIs
+
+- (void)setOutlineColor:(nonnull UIColor *)outlineColor forState:(MDCTextControlState)state {
+  [self.outlinedStyle setOutlineColor:outlineColor forState:state];
+  [self setNeedsLayout];
+}
+
+- (nonnull UIColor *)outlineColorForState:(MDCTextControlState)state {
+  return [self.outlinedStyle outlineColorForState:state];
+}
+
+#pragma mark Private Helpers
+
+- (MDCTextControlStyleOutlined *)outlinedStyle {
+  MDCTextControlStyleOutlined *outlinedStyle = nil;
+  if ([self.containerStyle isKindOfClass:[MDCTextControlStyleOutlined class]]) {
```

Comment:
Should we be throwing an NSAssert for the else clause of this? If the outlinedStyle is not a the class we are expecting that means we are doing something wrong?

Reply:
I can change this. There are a bunch of ways to prevent the MDCBaseTextField subclasses from setting borderStyle, I'm just looking for one that everyone will approve of. This is the mechanism currently in develop for doing it:
https://github.com/material-components/material-components-ios/blob/develop/components/TextFields/src/ContainedInputView/MDCBaseTextField.m#L260
It works, but before I merged it I received PR feedback to add the NSLog, which now prints every time an MDCOutlinedTextField is initialized, which is a problem. I can get rid of _all_ the borderStyle changes in this PR if I can just get rid of that log statement. Would you support that?

Reply:
^ that would simplify this PR and get rid of most of the changes that seem to be controversial.

Reply:
Never mind--marking borderStyle on filled and outlined textfields as unavailable, as discussed elsewhere in the PR.